src/app/icd/client/ICDRefreshKeyInfo.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "ICDRefreshKeyInfo.h"
+#include "CheckInDelegate.h"
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandPathParams.h>
+#include <app/InteractionModelEngine.h>
+#include <app/InteractionModelTimeout.h>
+#include <app/OperationalSessionSetup.h>
+#include <memory>
+
+namespace chip {
+namespace app {
+
+ICDRefreshKeyInfo::ICDRefreshKeyInfo(CheckInDelegate * apCheckInDelegate, ICDClientInfo * apICDClientInfo,
+                                     ICDClientStorage * aICDClientStorage) :
+    mOnConnectedCallback(HandleDeviceConnected, this),
+    mOnConnectionFailureCallback(HandleDeviceConnectionFailure, this)
+{
+    mpCheckInDelegate  = apCheckInDelegate;
+    mpICDClientInfo    = apICDClientInfo;
+    mpICDClientStorage = aICDClientStorage;
+}
+
+void RegisterCommandSenderDelegate::OnResponse(chip::app::CommandSender * apCommandSender,
+                                               const chip::app::ConcreteCommandPath & aPath, const chip::app::StatusIB & aStatus,
+                                               chip::TLV::TLVReader * aData)
+{
+    app::Clusters::IcdManagement::Commands::RegisterClientResponse::DecodableType responseData;
+    if (aData == nullptr)
+    {
+        mpICDRefreshKeyInfo->GetCheckInDelegate()->OnRegistrationUpdateFailure(mpICDRefreshKeyInfo.get(),
+                                                                               CHIP_ERROR_INVALID_TLV_ELEMENT);
+        Platform::Delete(this);
```

Comment:
You can create a helper function, for example, ReportError, then failure case in RegisterCommandSenderDelegate::OnResponse and  RegisterCommandSenderDelegate::OnError  can invoke the same code path for error processing